libs/src/services/creatorNode/CreatorNodeSelection.js

```diff
@@ -0,0 +1,153 @@
+    this.decisionTree.push({ stage: 'Get All Services', val: services })
+
+    if (this.whitelist && this.whitelist.length > 0) {
+      services = this.filterToWhitelist(services)
+      this.decisionTree.push({ stage: 'Filtered To Whitelist', val: services })
+    }
+
+    if (this.blacklist && this.blacklist.length > 0) {
+      services = this.filterFromBlacklist(services)
+      this.decisionTree.push({ stage: 'Filtered From Blacklist', val: services })
+    }
+
+    // Init map of creator node endpoints and their default stats
+    const map = {}
+    services.map(service => {
+      map[service] = {
+        version: null,
+        healthCheckEndpoint: `${service}/health_check`,
+        healthCheckResponseTime: null,
+        isHealthy: false,
+        isBehind: true,
+        isConfigured: false,
+        geographicData: {}
+        // todo: add other criteria we should evaluate like diskUsage, currentNumberOfRequests
+      }
+    })
+
+    const healthyEndpoints = []
+    const healthyServices = []
+    for (const service of services) {
+      const start = Date.now()
+      try {
+        // Check health and sync status, and then update stats per service
+        let healthCheckResp = await axios({
+          method: 'get',
+          url: map[service].healthCheckEndpoint
+        })
+        healthCheckResp = healthCheckResp.data.data
+
+        map[service].healthCheckResponseTime = Date.now() - start
+        map[service].version = healthCheckResp.version
+        const { isBehind, isConfigured } = await this.creatorNode.getSyncStatus(service)
+        map[service].isBehind = isBehind
+        map[service].isConfigured = isConfigured
+        map[service].geographicData = {
+          country: healthCheckResp.country,
+          latitude: healthCheckResp.latitude,
+          longitude: healthCheckResp.longitude
+        }
+
+        // If service is not configured, behind in blocks, or its major + minor version is behind, mark as unhealthy
+        map[service].isHealthy = isConfigured && !isBehind &&
+          this.ethContracts.hasSameMajorAndMinorVersion(this.currentVersion, healthCheckResp.version)
+
+        // Add to healthyServices list; used and sorted to select primary and secondaries
+        if (map[service].isHealthy) {
+          healthyEndpoints.push(service)
+          healthyServices.push({ endpoint: service, version: map[service].version, responseTime: map[service].healthCheckResponseTime })
+        }
+      } catch (e) {
+        console.error(`CreatorNodeSelection - Error with checking ${service} health: ${e}`)
+      }
+    }
+    this.decisionTree.push({ stage: 'Filtered Out Known Unhealthy', val: healthyEndpoints })
+
+    // Prioritize higher versions for selected creator nodes
+    this.sortBySemver(healthyServices)
+
+    // Index 0 of healthyServices will be the most optimal creator node candidate
+    const primary = healthyServices[0] ? healthyServices[0].endpoint : null
```

Comment:
if there's no healthy services is there just no primary?

Reply:
discussed offline -- will implement backup selecting logic